### ISSUE TITLE
Extremely slow inference speed

### ISSUE THREAD

**Issue by Thijsvandepoll** [pytorch/serve#2273]
### 🐛 Describe the bug

I am trying to deploy a PyTorch model retrieved from Huggingface (SentenceBert-like model) inside TorchServe in Docker. 

I have done the following: 

1. Create a `.mar` file of the model, moved it into a directory `model_store`,
2. Create a custom `config.properties` because I want to serve it in KServe later on. Note that inside here it exposes the model on port 8085.
3. Deploy the model inside docker using 
```
docker run --rm -it -p 8085:8085 -p 8082:8082 -v $(pwd)/model_store:/home/model-server/model-store
 -v $(pwd)/config.properties:/home/model-server/config.properties pytorch/torchserve:latest-cpu torchserve --start --model-store model-store --models model_name=model_name.mar --ncs
```
4. Call the model using a cURL request. This takes approximately 8 seconds to respond. I know that by just running a small python script, this model should respond within milliseconds. So this is extremely slow

As an alternative, I have tried building a custom Docker image with TorchServe. This worked, and responded with similar speeds as the normal python script. I have no idea what could be causing this difference in inference performance.

Some additional details:
- `transformers==4.25.1` is an dependency
- I have tried different versions of PyTorch and Transformers, and they all have the same problem, unless I create my own build.

### Error logs

Performance issue

### Installation instructions

The custom build that actually worked properly:

```
FROM python:3.9-bullseye

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
    gcc \
    libmariadb-dev \
    openjdk-11-jdk \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

RUN pip install \
      --disable-pip-version-check \
      --no-cache-dir \
      --no-compile \
      --upgrade \
      install transformers==4.25.1 torchtext==0.14.1 torchserve==0.7.1 torch==1.13.1 captum==0.6.0
```

### Model Packaing

torch-model-archiver

### config.properties

```
inference_address=http://0.0.0.0:8085
management_address=http://0.0.0.0:8085
metrics_address=http://0.0.0.0:8082
grpc_inference_port=7070
grpc_management_port=7071
enable_metrics_api=true
metrics_format=prometheus
enable_envvars_config=true
install_py_dep_per_model=true
model_store=/mnt/models/model-store
model_snapshot={"name":"startup.cfg","modelCount":1,"models":{"model_name":{"1.0":{"defaultVersion":true,"marName":"model_name.mar","minWorkers":1,"maxWorkers":5,"batchSize":1,"maxBatchDelay":10,"responseTimeout":120}}}}
```

### Versions

- pytorch/torchserve: 0.7.1
- torch: 1.13.1+cpu
- torchtext: 0.14.1
- transformers: 4.25.1

### Repro instructions

See description of the bug.

**Comment by Thijsvandepoll** [comment]
Thanks for the response! I ran the example again with the line removed and unfortunately that did not help. I was also not expecting it to work, because I have tested in my own Docker build with the exactly same configurations. And there is it quick.

Regarding the performance tuning with CPU inference, that I still need to do. Also I am planning on maybe serving it on GPU, but for now I was locally testing and found this big difference in performance. 

Any other ideas why this problem might occur? I could create the exact steps to follow to reproduce the problem if you want.

**Comment by msaroufim** [comment]
So just to be clear is the problem that you can observe much faster performance with your own custom dockerfile where you  run `pip install torchserve` vs using the official torchserve container?

**Comment by Thijsvandepoll** [comment]
Yeah correct, I am building a quick repo with a small README.md such that you can reproduce. Should be ready in 30 mins.

**Comment by Thijsvandepoll** [comment]
I have created the example repository with steps on how to reproduce. Let me know if you can reproduce it or if something is unclear:

https://github.com/Thijsvandepoll/torchserve-issue

**Comment by lxning** [comment]
@Thijsvandepoll could you use this tool [benchmark-ab.py](https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench) to run a benchmark to identify the high latency source?

**Comment by min-jean-cho** [comment]
Hi @Thijsvandepoll, could you modify your custom handler inference function to measure inference-only latency:

```python
def inference(self, data, *args, **kwargs):
  with torch.no_grad():
    marshalled_data = data.to(self.device)

    # warmup
    for _ in range(100):
        results = self.model(**marshalled_data, **kwargs)

    # measure
    import time
    start = time.time()
    for _ in range(100):
      results = self.model(**marshalled_data, **kwargs)
    end = time.time()
    print('Inference took {:.2f} ms in average'.format((end-start)/100*1000))
    
    return results
```

This helps to identify whether the performance difference is at the model inference-only level, or the end-to-end torchserve level.

**Comment by Thijsvandepoll** [comment]
@lxning I am really sorry for the slow response but a lot of stuff came in between :(. I am resuming my research into the slow responses now. 

I have run `ab` command for both models (i.e custom Docker image (port=8090) and Torchserve Docker image (port=8080)). 

The results for the **Torchserve Docker image** for `ab -T application/json -n 10 -c 1 -p example.json -k http://localhost:8080/predictions/custom_model` is:
```
Server Software:        
Server Hostname:        localhost
Server Port:            8080

Document Path:          /predictions/custom_model
Document Length:        17956 bytes

Concurrency Level:      1
Time taken for tests:   54.751 seconds
Complete requests:      10
Failed requests:        0
Keep-Alive requests:    10
Total transferred:      181910 bytes
Total body sent:        2580
HTML transferred:       179560 bytes
Requests per second:    0.18 [#/sec] (mean)
Time per request:       5475.069 [ms] (mean)
Time per request:       5475.069 [ms] (mean, across all concurrent requests)
Transfer rate:          3.24 [Kbytes/sec] received
                        0.05 kb/s sent
                        3.29 kb/s total

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       0
Processing:  5367 5475 119.6   5418    5705
Waiting:     5367 5475 119.5   5418    5705
Total:       5367 5475 119.6   5418    5705

Percentage of the requests served within a certain time (ms)
  50%   5418
  66%   5500
  75%   5531
  80%   5650
  90%   5705
  95%   5705
  98%   5705
  99%   5705
 100%   5705 (longest request)
```

The results for the **custom Docker image** for `ab -T application/json -n 10 -c 1 -p example.json -k http://localhost:8090/predictions/custom_model` is:
```
Server Software:        
Server Hostname:        localhost
Server Port:            8090

Document Path:          /predictions/custom_model
Document Length:        17959 bytes

Concurrency Level:      1
Time taken for tests:   0.803 seconds
Complete requests:      10
Failed requests:        0
Keep-Alive requests:    10
Total transferred:      181940 bytes
Total body sent:        2580
HTML transferred:       179590 bytes
Requests per second:    12.45 [#/sec] (mean)
Time per request:       80.349 [ms] (mean)
Time per request:       80.349 [ms] (mean, across all concurrent requests)
Transfer rate:          221.13 [Kbytes/sec] received
                        3.14 kb/s sent
                        224.26 kb/s total

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       0
Processing:    33   80 101.5     45     360
Waiting:       33   80 101.5     45     360
Total:         33   80 101.6     45     360

Percentage of the requests served within a certain time (ms)
  50%     45
  66%     46
  75%     48
  80%    119
  90%    360
  95%    360
  98%    360
  99%    360
 100%    360 (longest request)
```

From this benchmark it is clear that there is a difference. 


I am not entirely sure if `benchmark-ab.py` provides more insights than this, but I could not get it to work with my custom model. I tried to provide the path to the `.mar` file as per documentation that is possible, but it fails with an error that it cannot find the model. Also it fails to clean up it's resources keeping the ports occupied... If there is value in running `benchmark-ab.py`, could you point me on how to do that for the custom model?

**Comment by Thijsvandepoll** [comment]
@min-jean-cho Also sorry for this late response!

I have created the inference method like you suggested and deployed it on the custom Docker image vs on the Torchserve Docker image to compare inference times. Here are the results:

For the Torchserve Docker image (3 replications other timeout occurred):
```Inference took 5678.42 ms in average```

For the custom Docker image (100 replications):
```Inference took 39.45 ms in average```

Also here is a clear difference. Also I notice that on the custom image the CPU is fully being utilized while on the Torchserve image it is near to idle.

Some ideas what could cause this?

**Comment by Thijsvandepoll** [comment]
I think I found the cause of the issue. The problem seems to be related to my architecture. I am currently running the container on an Mac M2 chip. As shown above all runs ok, but slow. 

I ran the experiment on my old laptop having an Intel chip and there the performance seems similar between both images.

**Comment by Thijsvandepoll** [comment]
@ydennisy I build a custom torchserve image as shown above. That seems to be running normally. Not exactly sure why it works differently though.

**Comment by ppalmes** [comment]
it is possible that when you build docker from scratch, it is using the aarch architecture while pulling the docker image from official repo uses the intel architecture. mac has to use rosetta to translate it into arm architecture during runtime.

**Comment by GyuminJack** [comment]
hello. i experienced this situation too. in my opinion, it is not sure but i think jdk is different on 'pytorch/serve' and locally build image

openjdk-11 looks faster than openjdk-17 (torchserve image is using openjdk-17)
